Add support for decimal batch reader

wypb opened the pull request:
## Description
Add support for decimal batch reader

Benchmark(The lower the better)
```
Benchmark                                                          (byteArrayLength)  (decimalPrimitiveTypeName)  (enableOptimizedReader)  (nullable)  (writerVersion)  Mode  Cnt    Score    Error  Units
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A                      BINARY                     true        true      PARQUET_1_0  avgt   60   35.978 ±  0.213  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A                      BINARY                     true        true      PARQUET_2_0  avgt   60   56.711 ±  0.247  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A                      BINARY                     true       false      PARQUET_1_0  avgt   60   63.970 ±  0.239  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A                      BINARY                     true       false      PARQUET_2_0  avgt   60  104.495 ±  1.744  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A                      BINARY                    false        true      PARQUET_1_0  avgt   60   72.606 ±  0.766  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A                      BINARY                    false        true      PARQUET_2_0  avgt   60   64.700 ±  1.676  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A                      BINARY                    false       false      PARQUET_1_0  avgt   60  119.463 ±  2.066  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A                      BINARY                    false       false      PARQUET_2_0  avgt   60   97.458 ±  5.932  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A    FIXED_LEN_BYTE_ARRAY(16)                     true        true      PARQUET_1_0  avgt   60   19.863 ±  0.125  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A    FIXED_LEN_BYTE_ARRAY(16)                     true        true      PARQUET_2_0  avgt   60   57.809 ±  0.486  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A    FIXED_LEN_BYTE_ARRAY(16)                     true       false      PARQUET_1_0  avgt   60   29.744 ±  0.314  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A    FIXED_LEN_BYTE_ARRAY(16)                     true       false      PARQUET_2_0  avgt   60  100.839 ±  0.838  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A    FIXED_LEN_BYTE_ARRAY(16)                    false        true      PARQUET_1_0  avgt   60   69.015 ±  0.418  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A    FIXED_LEN_BYTE_ARRAY(16)                    false        true      PARQUET_2_0  avgt   60   67.082 ±  2.169  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A    FIXED_LEN_BYTE_ARRAY(16)                    false       false      PARQUET_1_0  avgt   60  113.877 ±  3.237  ns/op
BenchmarkDecimalColumnBatchReader.readLongDecimal                                N/A    FIXED_LEN_BYTE_ARRAY(16)                    false       false      PARQUET_2_0  avgt   60  101.167 ±  4.742  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT32                     true        true      PARQUET_1_0  avgt   60    5.892 ±  0.081  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT32                     true        true      PARQUET_2_0  avgt   60    7.909 ±  0.077  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT32                     true       false      PARQUET_1_0  avgt   60    2.198 ±  0.016  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT32                     true       false      PARQUET_2_0  avgt   60    6.230 ±  0.190  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT32                    false        true      PARQUET_1_0  avgt   60   25.590 ±  2.622  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT32                    false        true      PARQUET_2_0  avgt   60   20.364 ±  0.143  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT32                    false       false      PARQUET_1_0  avgt   60   16.077 ±  0.222  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT32                    false       false      PARQUET_2_0  avgt   60   20.276 ±  3.783  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT64                     true        true      PARQUET_1_0  avgt   60   13.379 ±  4.040  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT64                     true        true      PARQUET_2_0  avgt   60    9.310 ±  0.036  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT64                     true       false      PARQUET_1_0  avgt   60    3.766 ±  0.035  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT64                     true       false      PARQUET_2_0  avgt   60    8.813 ±  0.083  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT64                    false        true      PARQUET_1_0  avgt   60   25.213 ±  0.222  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT64                    false        true      PARQUET_2_0  avgt   60   21.266 ±  0.068  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT64                    false       false      PARQUET_1_0  avgt   60   22.471 ±  0.617  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                       INT64                    false       false      PARQUET_2_0  avgt   60   18.181 ±  0.937  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                      BINARY                     true        true      PARQUET_1_0  avgt   60   13.513 ±  1.030  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                      BINARY                     true        true      PARQUET_2_0  avgt   60   31.413 ±  7.525  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                      BINARY                     true       false      PARQUET_1_0  avgt   60   16.384 ±  2.283  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                      BINARY                     true       false      PARQUET_2_0  avgt   60   37.136 ±  0.573  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                      BINARY                    false        true      PARQUET_1_0  avgt   60   38.115 ±  2.961  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                      BINARY                    false        true      PARQUET_2_0  avgt   60   33.686 ±  0.890  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                      BINARY                    false       false      PARQUET_1_0  avgt   60   38.408 ±  0.306  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A                      BINARY                    false       false      PARQUET_2_0  avgt   60   39.839 ±  0.300  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A     FIXED_LEN_BYTE_ARRAY(8)                     true        true      PARQUET_1_0  avgt   60    6.254 ±  0.055  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A     FIXED_LEN_BYTE_ARRAY(8)                     true        true      PARQUET_2_0  avgt   60   18.820 ±  0.159  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A     FIXED_LEN_BYTE_ARRAY(8)                     true       false      PARQUET_1_0  avgt   60    3.435 ±  0.110  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A     FIXED_LEN_BYTE_ARRAY(8)                     true       false      PARQUET_2_0  avgt   60   29.901 ±  0.606  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A     FIXED_LEN_BYTE_ARRAY(8)                    false        true      PARQUET_1_0  avgt   60   31.011 ±  0.310  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A     FIXED_LEN_BYTE_ARRAY(8)                    false        true      PARQUET_2_0  avgt   60   35.836 ±  3.396  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A     FIXED_LEN_BYTE_ARRAY(8)                    false       false      PARQUET_1_0  avgt   60   36.312 ±  0.709  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimal                               N/A     FIXED_LEN_BYTE_ARRAY(8)                    false       false      PARQUET_2_0  avgt   60   61.468 ± 15.297  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  1                         N/A                     true        true      PARQUET_1_0  avgt   60    0.784 ±  0.046  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  1                         N/A                     true        true      PARQUET_2_0  avgt   60    6.970 ±  0.188  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  1                         N/A                     true       false      PARQUET_1_0  avgt   60    0.842 ±  0.048  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  1                         N/A                     true       false      PARQUET_2_0  avgt   60   12.153 ±  5.229  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  1                         N/A                    false        true      PARQUET_1_0  avgt   60   39.380 ±  4.667  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  1                         N/A                    false        true      PARQUET_2_0  avgt   60   21.779 ±  1.257  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  1                         N/A                    false       false      PARQUET_1_0  avgt   60   33.686 ±  0.484  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  1                         N/A                    false       false      PARQUET_2_0  avgt   60   18.937 ±  0.199  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  2                         N/A                     true        true      PARQUET_1_0  avgt   60    1.316 ±  0.030  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  2                         N/A                     true        true      PARQUET_2_0  avgt   60   27.600 ±  0.792  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  2                         N/A                     true       false      PARQUET_1_0  avgt   60    1.307 ±  0.032  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  2                         N/A                     true       false      PARQUET_2_0  avgt   60   27.122 ±  0.425  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  2                         N/A                    false        true      PARQUET_1_0  avgt   60   33.514 ±  0.238  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  2                         N/A                    false        true      PARQUET_2_0  avgt   60   43.301 ±  1.226  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  2                         N/A                    false       false      PARQUET_1_0  avgt   60   33.281 ±  0.459  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  2                         N/A                    false       false      PARQUET_2_0  avgt   60   39.255 ±  0.485  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  3                         N/A                     true        true      PARQUET_1_0  avgt   60    1.710 ±  0.022  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  3                         N/A                     true        true      PARQUET_2_0  avgt   60   28.296 ±  0.558  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  3                         N/A                     true       false      PARQUET_1_0  avgt   60    1.813 ±  0.094  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  3                         N/A                     true       false      PARQUET_2_0  avgt   60   27.826 ±  0.314  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  3                         N/A                    false        true      PARQUET_1_0  avgt   60   36.132 ±  1.231  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  3                         N/A                    false        true      PARQUET_2_0  avgt   60   41.832 ±  1.558  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  3                         N/A                    false       false      PARQUET_1_0  avgt   60   34.004 ±  0.396  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  3                         N/A                    false       false      PARQUET_2_0  avgt   60   38.750 ±  0.302  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  4                         N/A                     true        true      PARQUET_1_0  avgt   60    2.203 ±  0.043  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  4                         N/A                     true        true      PARQUET_2_0  avgt   60   28.282 ±  0.533  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  4                         N/A                     true       false      PARQUET_1_0  avgt   60    2.149 ±  0.029  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  4                         N/A                     true       false      PARQUET_2_0  avgt   60   28.413 ±  0.598  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  4                         N/A                    false        true      PARQUET_1_0  avgt   60   34.270 ±  0.404  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  4                         N/A                    false        true      PARQUET_2_0  avgt   60   39.215 ±  0.549  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  4                         N/A                    false       false      PARQUET_1_0  avgt   60   36.183 ±  0.952  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  4                         N/A                    false       false      PARQUET_2_0  avgt   60   40.536 ±  0.995  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  5                         N/A                     true        true      PARQUET_1_0  avgt   60    2.498 ±  0.101  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  5                         N/A                     true        true      PARQUET_2_0  avgt   60   28.847 ±  0.456  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  5                         N/A                     true       false      PARQUET_1_0  avgt   60    2.499 ±  0.083  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  5                         N/A                     true       false      PARQUET_2_0  avgt   60   29.880 ±  0.273  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  5                         N/A                    false        true      PARQUET_1_0  avgt   60   34.867 ±  0.772  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  5                         N/A                    false        true      PARQUET_2_0  avgt   60   41.475 ±  1.487  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  5                         N/A                    false       false      PARQUET_1_0  avgt   60   38.280 ±  4.063  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  5                         N/A                    false       false      PARQUET_2_0  avgt   60   39.719 ±  0.523  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  6                         N/A                     true        true      PARQUET_1_0  avgt   60    2.642 ±  0.061  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  6                         N/A                     true        true      PARQUET_2_0  avgt   60   29.860 ±  0.367  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  6                         N/A                     true       false      PARQUET_1_0  avgt   60    2.663 ±  0.042  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  6                         N/A                     true       false      PARQUET_2_0  avgt   60   29.849 ±  0.626  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  6                         N/A                    false        true      PARQUET_1_0  avgt   60   35.169 ±  0.158  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  6                         N/A                    false        true      PARQUET_2_0  avgt   60   40.821 ±  0.510  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  6                         N/A                    false       false      PARQUET_1_0  avgt   60   38.378 ±  1.814  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  6                         N/A                    false       false      PARQUET_2_0  avgt   60   44.101 ±  4.470  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  7                         N/A                     true        true      PARQUET_1_0  avgt   60    2.788 ±  0.036  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  7                         N/A                     true        true      PARQUET_2_0  avgt   60   29.854 ±  0.341  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  7                         N/A                     true       false      PARQUET_1_0  avgt   60    2.868 ±  0.052  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  7                         N/A                     true       false      PARQUET_2_0  avgt   60   30.821 ±  0.544  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  7                         N/A                    false        true      PARQUET_1_0  avgt   60   38.564 ±  1.608  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  7                         N/A                    false        true      PARQUET_2_0  avgt   60   41.938 ±  0.737  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  7                         N/A                    false       false      PARQUET_1_0  avgt   60   37.818 ±  2.376  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  7                         N/A                    false       false      PARQUET_2_0  avgt   60   86.133 ± 29.822  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  8                         N/A                     true        true      PARQUET_1_0  avgt   60    4.066 ±  0.423  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  8                         N/A                     true        true      PARQUET_2_0  avgt   60   34.223 ±  3.815  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  8                         N/A                     true       false      PARQUET_1_0  avgt   60    4.593 ±  0.343  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  8                         N/A                     true       false      PARQUET_2_0  avgt   60   30.767 ±  1.049  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  8                         N/A                    false        true      PARQUET_1_0  avgt   60   50.825 ±  7.990  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  8                         N/A                    false        true      PARQUET_2_0  avgt   60   55.074 ±  6.835  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  8                         N/A                    false       false      PARQUET_1_0  avgt   60   43.135 ±  9.506  ns/op
BenchmarkDecimalColumnBatchReader.readShortDecimalByteArrayLength                  8                         N/A                    false       false      PARQUET_2_0  avgt   60   48.081 ±  5.818  ns/op
```

## Impact
When we enable Parquet batch reader(`parquet_batch_read_optimization_enabled`=true), the Decimal type will read data in Batch mode.

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes

```
== RELEASE NOTES ==

Hive Connector Changes
* Add support for decimal batch reader :pr:`22636`
```

CC: @zhenxiao

yingsu00 reviewed (commented):
@wypb Thank you so much for this great work. I wonder if you could add tests?

wypb commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/plain/ShortDecimalPlainValuesDecoder.java:
```diff
@@ -0,0 +1,92 @@
+        // Equivalent to expectedValue = bytes[endOffset] < 0 ? -1 : 0
+        byte expectedValue = (byte) (bytes[endOffset] >> 7);
+        for (int i = offset; i < endOffset; i++) {
+            if (bytes[i] != expectedValue) {
```
The bytes[endOffset] of the above code is actually the most significant byte. To illustrate this better, I built a test locally. The original data is 123456789012.12345678, typeLength=14.  the contents of `bytes` are [2, 0, 0, 0, 3, 1, 0, 0, 0, 0, 0, 0, -85, 84, -87, -116, -23, -53, -11, 78], In this scenario, inputBytesOffset = 6, extraBytesLength = 6, so endOffset = 12, bytes[endOffset] = -85, which is actually the most significant byte. Then we check whether bytes[6.. 11] is -1. Attached is the file I tested.
[d93db016-5b4d-4d3e-baec-be116039c66f.tgz](https://github.com/prestodb/presto/files/15260224/d93db016-5b4d-4d3e-baec-be116039c66f.tgz)

wypb commented on presto-parquet/src/main/java/com/facebook/presto/parquet/ParquetTypeUtils.java:
```diff
@@ -256,17 +259,38 @@ public static boolean isValueNull(boolean required, int definitionLevel, int max
+        switch (length) {
+            case 8:
+                value |= bytes[startOffset + 7] & 0xFFL;
+                // fall through
```
Fixed.

wypb commented on presto-parquet/src/main/java/com/facebook/presto/parquet/ParquetTypeUtils.java:
```diff
@@ -335,4 +359,31 @@ public static boolean isTimeStampMicrosType(ColumnDescriptor descriptor)
+        }
+
+        DecimalLogicalTypeAnnotation decimalLogicalTypeAnnotation = (DecimalLogicalTypeAnnotation) logicalTypeAnnotation;
+        return decimalLogicalTypeAnnotation.getPrecision() <= Decimals.MAX_SHORT_PRECISION;
```
Fixed.

wypb commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/ShortDecimalFixedWidthByteArrayBatchDecoder.java:
```diff
@@ -0,0 +1,274 @@
+public class ShortDecimalFixedWidthByteArrayBatchDecoder
+{
+    private static final ShortDecimalDecoder[] VALUE_DECODERS = new ShortDecimalDecoder[] {
+            new BigEndianReader1(),
```
This actually further optimizes the reading speed of short decimals. The implementation of `ShortDecimalFixedWidthByteArrayBatchDecoder` actually refers to the implementation of Trino: https://github.com/trinodb/trino/commit/f71a8156db0d5810b199ade93f3affca650b8ba8

elharo commented on presto-parquet/src/main/java/com/facebook/presto/parquet/ParquetTypeUtils.java:
```diff
@@ -256,17 +259,31 @@ public static boolean isValueNull(boolean required, int definitionLevel, int max
     // copied from presto-hive DecimalUtils
```
is this still true?

wypb replied:
Good catch, thank you @elharo I've deleted it.

elharo commented on presto-parquet/src/main/java/com/facebook/presto/parquet/ParquetTypeUtils.java:
```diff
@@ -256,17 +259,31 @@ public static boolean isValueNull(boolean required, int definitionLevel, int max
 
-        for (int i = 0; i < bytes.length; i++) {
-            value |= ((long) bytes[bytes.length - i - 1] & 0xFFL) << (8 * i);
+    public static long getShortDecimalValue(byte[] bytes, int startOffset, int length)
```
can this be private or not public?

wypb replied:
No, this method is used in many places.

wypb commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/delta/LongDecimalApacheParquetValueDecoder.java:
```diff
@@ -0,0 +1,62 @@
+import static io.airlift.slice.SizeOf.SIZE_OF_LONG;
+import static java.util.Objects.requireNonNull;
+
+public class LongDecimalApacheParquetValueDecoder
```
I have renamed `LongDecimalApacheParquetValueDecoder` and `ShortDecimalApacheParquetValueDecoder` to `FixedLenByteArrayShortDecimalDeltaValueDecoder` and `FixedLenByteArrayLongDecimalDeltaValueDecoder` respectively.

elharo commented on presto-parquet/src/test/java/com/facebook/presto/parquet/reader/TestData.java:
```diff
@@ -0,0 +1,199 @@
+        return data;
+    }
+
+    private static int propagateSignBit(int value, int bitsToPad)
```
why here and in ByteUtils?

wypb replied:
I looked at the code and found that this can actually be deleted.

wypb commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/delta/Int64ShortDecimalDeltaValuesDecoder.java:
```diff
@@ -0,0 +1,55 @@
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class Int64ShortDecimalDeltaValuesDecoder
```
The common parts of `Int32ShortDecimalDeltaValuesDecoder` and `Int64ShortDecimalDeltaValuesDecoder` have been extracted into `AbstractInt64AndInt32ShortDecimalDeltaValuesDecoder`.

wypb commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/delta/BinaryShortDecimalDeltaValuesDecoder.java:
```diff
@@ -0,0 +1,74 @@
+import static com.facebook.presto.parquet.ParquetTypeUtils.getShortDecimalValue;
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class BinaryShortDecimalDeltaValuesDecoder
```
I'll see how to refactor these two classes because they implement different interfaces.

elharo commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/BytesUtils.java:
```diff
@@ -61,4 +61,9 @@ public static void unpack8Values(byte inByte, byte[] out, int outPos)
         out[7 + outPos] = (byte) (inByte >> 7 & 1);
     }
+
+    public static long propagateSignBit(long value, int bitsToPad)
```
This method could use a unit test that addresses it directly

elharo commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/SimpleSliceInputStream.java:
```diff
@@ -0,0 +1,167 @@
+    public SimpleSliceInputStream(Slice slice, int offset)
+    {
+        this.slice = requireNonNull(slice, "slice is null");
+        checkArgument(slice.length() == 0 || slice.hasByteArray(), "SimpleSliceInputStream supports only slices backed by byte array");
```
only supports

wypb replied:
Fixed.

wypb commented on presto-common/src/main/java/com/facebook/presto/common/type/UnscaledDecimal128Arithmetic.java:
```diff
@@ -1685,7 +1685,7 @@ else if (high != 0x4b3b4ca85a86c47aL) {
     }
 
-    private static byte[] reverse(final byte[] a)
+    public static byte[] reverse(final byte[] a)
```
The first version of the code used this method. Later I refactored the code, so this method is indeed not used now.

wypb commented on presto-parquet/src/test/java/com/facebook/presto/parquet/BenchmarkParquetReader.java:
```diff
@@ -94,7 +94,13 @@ public static void main(String[] args)
+    public static Object read(BenchmarkData data, boolean enableOptimizedReader)
+            throws Exception
+    {
+        try (ParquetReader recordReader = data.createRecordReader(enableOptimizedReader)) {
```
Already refactored.

wypb commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/Decoders.java:
```diff
@@ -110,8 +134,25 @@ private static final ValuesDecoder createValuesDecoder(ColumnDescriptor columnDe
+                    if (prestoType.isPresent()) {
+                        DecimalType decimalType = (DecimalType) prestoType.get();
+                        if (decimalType.isShort()) {
+                            return new ShortDecimalPlainValuesDecoder(columnDescriptor, buffer, offset, length);
```
Already refactored.

wypb commented on presto-parquet/src/main/java/com/facebook/presto/parquet/ColumnReaderFactory.java:
```diff
@@ -58,25 +60,43 @@ private ColumnReaderFactory()
                     if (isTimeStampMicrosType(descriptor)) {
                         return isNested ? new Int64TimestampMicrosNestedBatchReader(descriptor) : new Int64TimestampMicrosFlatBatchReader(descriptor);
                     }
+
```
The following code has been added
```
int precision = ((DecimalLogicalTypeAnnotation) descriptor.getPrimitiveType().getLogicalTypeAnnotation()).getPrecision();
if (precision < 10) {
    log.warn("PrimitiveTypeName is INT64 but precision is less then 10.");
}
```

elharo reviewed (commented):
This PR seems short on new tests. I would have expected quite a number given all the new code.

wypb commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/plain/ShortDecimalPlainValuesDecoder.java:
```diff
@@ -0,0 +1,92 @@
+        int inputBytesOffset = input.getByteArrayOffset();
+        for (int i = offset; i < offset + length; i++) {
+            checkBytesFitInShortDecimal(inputBytes, inputBytesOffset, extraBytesLength, columnDescriptor);
+            values[i] = getShortDecimalValue(inputBytes, inputBytesOffset + extraBytesLength, Long.BYTES);
```
According to the implementation of the code, `extraBytesLength` will not be less than 0, but must be greater than 0.
```
if (typeLength <= Long.BYTES) {
    ....        
}
int extraBytesLength = typeLength - Long.BYTES;
```

wypb commented:
@zhenxiao @yingsu00 @elharo sorry for the late reply. I have made modifications according to the previous review comments.

> I wonder if you could add tests?

I generated Parquet files with different encodings locally. I will see how to add them to the test set.
In addition, `BenchmarkShortDecimalColumnReader` and `BenchmarkLongDecimalColumnReader` have different coding data verification logic in batch mode and non-batch mode.

steveburnett commented:
Nit, suggest release note entry change: 

```
== RELEASE NOTES ==

Hive Connector Changes
* Add support decimal batch reader :pr:`22636`
```

wypb commented:
@zhenxiao @yingsu00 I added three new tests in `AbstractTestParquetReader.java`: `testLongDecimalBackedByBinary`, `testShortDecimalBackedByBinary` and `testShortDecimalBackedByFixedLenByteArray`, plus the previous `testDecimalBackedByFixedLenByteArray` (already named to `testLongDecimalBackedByFixedLenByteArray`), `testDecimalBackedByINT64` and `DecimalBackedByINT32` test just covers all the Decoder I added this time

wypb commented:
@zhenxiao @yingsu00 Any comments on this?

elharo reviewed (commented):
There's a lot of new public API here that feels like it needs unit tests, e.g. SimpleSliceInputStream

tdcmeehan commented:
@wypb could you add more tests to address @elharo's feedback?

wypb commented:
HI @tdcmeehan @elharo Thank you for your review, I'll add some test cases soon.

yingsu00 commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/Decoders.java:
```diff
@@ -173,11 +234,46 @@ private static final ValuesDecoder createValuesDecoder(ColumnDescriptor columnDe
         if ((encoding == DELTA_BYTE_ARRAY || encoding == DELTA_LENGTH_BYTE_ARRAY) && type == PrimitiveTypeName.BINARY) {
             ByteBufferInputStream inputStream = ByteBufferInputStream.wrap(ByteBuffer.wrap(buffer, offset, length));
+
+            Optional<Type> prestoType = createDecimalType(columnDescriptor);
```
I know this is from existing code, but it is confusing to detect if the logical type is decimal by calling createDecimalType() and check its returned type. It'll be clearer to make the type detecting consistent with other encodings, e.g. as what you did on line 195, 196

wypb replied:
Already refactored, thanks.

yingsu00 commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/plain/FixedLenByteArrayShortDecimalPlainValuesDecoder.java:
```diff
@@ -0,0 +1,91 @@
+
+    public FixedLenByteArrayShortDecimalPlainValuesDecoder(ColumnDescriptor columnDescriptor, byte[] byteBuffer, int bufferOffset, int length)
+    {
+        this.columnDescriptor = columnDescriptor;
```
I know the existing decoders code has the same pattern, but can you please add requireNonNull() on columnDescriptor and byteBuffer? Same for all other constructors for the classes you add. Thanks!

yingsu00 reviewed (commented):
@wypb Will you please add the tests @elharo requested? For example, you can reference TestBooleanStream for writing tests for SimpleSliceInputStream. Thank you!

wypb commented:
HI @elharo @tdcmeehan @yingsu00 thanks for your inputs and sorry for the delay. 

I modified some code according to the previous review and added a new test case `com.facebook.presto.parquet.batchreader.TestSimpleSliceInputStream`. I have rebased the code and the diff file can be found in https://github.com/prestodb/presto/commit/d527a91a5ecafe8c88914639d22444119dbd5d62.

elharo commented on presto-hive/src/test/java/com/facebook/presto/hive/parquet/AbstractTestParquetReader.java:
```diff
@@ -938,24 +939,63 @@ public void testDecimalBackedByINT64()
+            throws Exception
+    {
+        for (int precision = 1; precision <= MAX_SHORT_PRECISION; precision++) {
+            int scale = ThreadLocalRandom.current().nextInt(precision);
```
I get very nervous seeing random in tests. If something fails it's not reproducible, Use fixed constant test data instead, even if the fixed values are initially randomly chosen.

elharo commented on presto-hive/src/test/java/com/facebook/presto/hive/parquet/AbstractTestParquetReader.java:
```diff
@@ -938,24 +939,63 @@ public void testDecimalBackedByINT64()
+            throws Exception
+    {
+        for (int precision = 1; precision <= MAX_SHORT_PRECISION; precision++) {
+            int scale = ThreadLocalRandom.current().nextInt(precision);
```
ditto

elharo commented on presto-parquet/src/main/java/com/facebook/presto/parquet/ColumnReaderFactory.java:
```diff
@@ -40,43 +41,67 @@
+                    if (!isNested && isShortDecimalType(descriptor)) {
+                        int precision = ((DecimalLogicalTypeAnnotation) descriptor.getPrimitiveType().getLogicalTypeAnnotation()).getPrecision();
+                        if (precision < 10) {
+                            log.warn("PrimitiveTypeName is INT64 but precision is less then 10.");
```
delete the warning; it's not actionable

or if I'm wrong and this is a real problem, then a waring isn't enough. This should fail outright

elharo commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/delta/FixedLenByteArrayShortDecimalDeltaValueDecoder.java:
```diff
@@ -0,0 +1,84 @@
+ * is not a common one, just use the existing one provided by Parquet library and add a wrapper around it that satisfies the
+ * {@link ShortDecimalValuesDecoder} interface.
+ */
+public class FixedLenByteArrayShortDecimalDeltaValueDecoder
```
avoid abbreviations; thus FixedLen --> FixedLength

but aren't all arrays in java fixed length? So maybe just ByteArrayShortDecimalDeltaValueDecoder

wypb replied:
The naming convention of the Decoder class name here is ParquetPrimitiveType + [Short|Long]Decimal + encoding + ValuesDecoder. The PrimitiveTypeName corresponding to this class is FIXED_LEN_BYTE_ARRAY, so this name is used.

elharo commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/plain/BinaryLongDecimalPlainValuesDecoder.java:
```diff
@@ -0,0 +1,73 @@
+            int positionOffset = offsets[i];
+            int positionLength = offsets[i + 1] - positionOffset;
+            byte[] temp = new byte[positionLength];
+            System.arraycopy(byteBuffer, positionOffset, temp, 0, positionLength);
```
Personally I find Arrays.copyOf a little easier to read; up to you

wypb replied:
I changed it to `byte[] temp = Arrays.copyOfRange(byteBuffer, positionOffset, positionOffset + positionLength);`

elharo commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/plain/FixedLenByteArrayLongDecimalPlainValuesDecoder.java:
```diff
@@ -0,0 +1,82 @@
+    @Override
+    public void readNext(long[] values, int offset, int length)
+    {
+        final byte[] localByteBuffer = byteBuffer;
```
Danger! Since this local variable is just a reference, byteBuffer is modified anyway when localByteBuffer is. Code below might be correct, I'm not sure, but this variable should be removed.

wypb replied:
Nice catch.

elharo commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/plain/FixedLenByteArrayShortDecimalPlainValuesDecoder.java:
```diff
@@ -0,0 +1,92 @@
+        byte expectedValue = (byte) (bytes[endOffset] >> 7);
+        for (int i = offset; i < endOffset; i++) {
+            if (bytes[i] != expectedValue) {
+                throw new PrestoException(NOT_SUPPORTED, format(
```
string concatenation is simpler here

wypb replied:
Fixed.

elharo commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/plain/ShortDecimalFixedWidthByteArrayBatchDecoder.java:
```diff
@@ -0,0 +1,274 @@
+
+                // We first shift the byte as left as possible. Then, when shifting back right,
+                // the sign bit will get propagated
+                values[offset] = value << 56 >> 56;
```
Order of operations is foggy. Please use parentheses to make this explicit.

wypb replied:
Fixed.

elharo commented on presto-parquet/src/test/java/com/facebook/presto/parquet/reader/TestData.java:
```diff
@@ -0,0 +1,93 @@
+            long[] result = new long[size];
+            for (int i = 0; i < size; i++) {
+                result[i] = Math.max(
+                        Math.min(randomLong(random, bitWidth), max),
```
avoid random numbers in tests. Test results need to be reproducible.

wypb replied:
Fixed.

wypb commented:
Thank you @elharo for the review! I have updated with a new (squashed) commit, and I have addressed all your comments. Let me know if I missed anything.

yingsu00 commented:
@wypb There is a test failure: Failures: 
[ERROR]   TestPrestoNativeIcebergTpcdsQueriesParquetUsingThrift.doDeletesAndQuery:399->verifyDeletes:390->AbstractTestQueryFramework.computeScalar:149->AbstractTestQueryFramework.computeActual:139 ? Runtime size <= capacity_ (18446744073709551176 vs. 1440) Split [Hive: file:/tmp/iceberg_data/HIVE/tpcds/store_sales/data/183baea9-babf-4daa-875b-2148de383c5a.parquet 0 - 2071914] Task 20240710_023723_00114_2m4dj.1.0.2.0 Operator: TableScan[0] 0

Can you please investigate? THanks!

tdcmeehan commented:
@yingsu00 this is actually likely due to https://github.com/facebookincubator/velox/pull/10261 and is being backed out in  https://github.com/facebookincubator/velox/pull/10431.  It was added in https://github.com/prestodb/presto/pull/23138 (which appeared to have been merged in spite of the failure it introduced).  See: https://github.com/prestodb/presto/issues/23156

wypb commented:
Hi @tdcmeehan  thank you for sharing the information. @yingsu00 I synchronized the latest code, and now CI is all green.

yingsu00 commented:
> @yingsu00 this is actually likely due to [facebookincubator/velox#10261](https://github.com/facebookincubator/velox/pull/10261) and is being backed out in [facebookincubator/velox#10431](https://github.com/facebookincubator/velox/pull/10431). It was added in #23138 (which appeared to have been merged in spite of the failure it introduced). See: #23156

Thank you @tdcmeehan for letting me know!
